### DISCUSSION TITLE
Uninstallable 1.8.7 patch releases

### DISCUSSION THREAD

**Discussion by thegoodlittlescript** [rbenv/ruby-build#2255] [Build failures]
When we try to install ruby-1.8.7-p375 we get:

```shell
rbenv install 1.8.7-p375
tail -f /tmp/ruby-build.....log
# warning: Could not find remote branch dabc90361d1b71fcac4f606eb1eb1fa25c047fd9 to clone.
# fatal: Remote branch dabc90361d1b71fcac4f606eb1eb1fa25c047fd9 not found in upstream origin
```

This error lines up with changes in https://github.com/rbenv/ruby-build/pull/2120 where the 1.8.7-p375 config was updated to reference the dabc9036 commit.

```
install_git "ruby-1.8.7-p375" "https://github.com/ruby/ruby.git" "dabc90361d1b71fcac4f606eb1eb1fa25c047fd9" warn_eol autoconf auto_tcltk standard
```

So it looks like the error means install_git doesn't support a commit. Notably a branch/tag works. I updated the commit to an existing branch (`ruby_1_8_7`) and later a tag (`v1_8_7_374`). I was able to successfully install with both. So ultimately the config needs a tag corresponding to 1.8.7-p375. Problem is no such tag exists in [ruby/ruby](https://github.com/ruby/ruby). The latest is p374.

![Screenshot 2023-09-19 at 9 02 14 AM](https://github.com/rbenv/ruby-build/assets/51981126/5c1b23b3-03ef-4836-8f97-609059f43f9a)

Likewise the latest release of 1.8.7 is p374 on https://www.ruby-lang.org/en/downloads/releases/:

![Screenshot 2023-09-19 at 10 12 35 AM](https://github.com/rbenv/ruby-build/assets/51981126/6483dc85-67ac-4a05-8d0f-19dc58c4b100)

Where does 1.8.7-p375 come from then? Turns out there is a commit on the [ruby_1_8_7 branch](https://github.com/ruby/ruby/commits/ruby_1_8_7) corresponding to 1.8.7-p375 ([dabc9036](https://github.com/ruby/ruby/commit/dabc90361d1b71fcac4f606eb1eb1fa25c047fd9), same one as above). There's also a commit for for 1.8.7-p376 ([f48ae0d1](https://github.com/ruby/ruby/commit/f48ae0d10c5b586db5748b0d4b645c7e9ff5d52e)). Looks like commits were made for each of these patch versions, but tags / releases were not made. Ultimately this prevents the 1.8.7-p375 install.

Possible solutions:

- Make tags in ruby/ruby. Update the config files to use the tags.
- Update `install_git` to work off of a commit

Observed using:

- Debian GNU/Linux 8 (jessie)
- `rbenv 1.2.0`
- `ruby-build 20230914.1`
- `git version 2.1.4`

**Comment by mislav** [comment]
Thank you for researching and pointing this out! Note that since this is a legitimate bug with how `install_git` works, this could have been a regular Issue rather than a Discussion. But it doesn't matter so much.

> Possible solutions:
> 
> * Make tags in ruby/ruby. Update the config files to use the tags.
> * Update `install_git` to work off of a commit

Ruby 1.8 is so old that I doubt that the Ruby core team would be interested in publishing tags for ancient versions. I would rather just fix `install_git` to work with a commit SHA instead of a branch name.

The problem is here:
https://github.com/rbenv/ruby-build/blob/bfffe34ba100ae84e5e7bde1a10bf01707e18c05/bin/ruby-build#L493

We could detect the length of the git ref given to `install_git` and, if it's 40 characters, assume that it's a commit SHA rather than a branch name. The downside is that this approach could break anyone using a tag/branch name that just so happens to be 40 characters long. Thoughts?

**Reply by thegoodlittlescript** [reply]
Hi @mislav thanks! Good to know. I started this as an issue but misinterpreted the instructions and put it here.

Your solution sounds reasonable and unlikely to cause problems in practice. I like that it doesn't require Ruby core to take action ... as you said ruby 1.8.7 is very old. If the possible collision ends up bugging you, then as an alternative I think that line could be rewritten. 

Looks to me like the code is:

- if package_name exists, then fetch/checkout 
- else clone (bug here)

This might be rewritten as:

- if package_name doesn't exist, then init an empty repo
- fetch/reset

Reason being fetch already supports branch/tag/commit and you could use FETCH_HEAD to checkout whatever comes back. Or at least that's how it seems. It's a bit unusual. Here's a demo:

```
#!/bin/bash
fetch_git_demo() {
  local package_name="$1"
  local git_url="$2"
  local git_ref="$3"

  echo "Cloning ${git_url}... ($package_name)" >&2

  if ! [ -e "${package_name}" ]
  then
    mkdir "${package_name}"
    (cd "${package_name}"; git init)
  fi
  ( cd "${package_name}"
    git fetch --depth 1 "$git_url" "+${git_ref}"
    git reset --hard FETCH_HEAD
  ) 
}
fetch_git_demo clone-by-sha "https://github.com/ruby/ruby.git" "dabc90361d1b71fcac4f606eb1eb1fa25c047fd9"
fetch_git_demo clone-by-branch "https://github.com/ruby/ruby.git" "ruby_1_8_7"  # resolves to f48ae0d10c5b586db5748b0d4b645c7e9ff5d52e
fetch_git_demo clone-by-tag "https://github.com/ruby/ruby.git" "v1_8_7_374"     # resolves to 83c8cd7199f25dfb490934a5227ec26770e292e4
```

ps I've been out of ruby for a while and we've never met, but I remember you from will_paginate. Great to run into you online! Thank you for all your great work over the years.
